crates/sim2h/src/naive_sharding.rs

```diff
@@ -0,0 +1,131 @@
+            .unwrap()
+        }
+    }
+}
+
+/// implement a super simple sharding algorithm
+/// to distribute data when node counts go > 50
+pub fn naive_sharding_should_store(
+    agent_loc: Location,
+    data_addr_loc: Location,
+    node_count: u64,
+) -> bool {
+    // if there are < 50 nodes, everyone should store everything
+    if node_count < 50 {
+        return true;
+    }
+
+    // divide up the space so on average data will be stored by 50 nodes
+    let dist: f64 = 4294967295.0 / (node_count as f64 / 50.0);
+
+    // determine if this specific piece of data should be stored by this node
+    agent_loc.forward_distance_to(data_addr_loc) < dist as u32
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use lib3h_sodium::SodiumCryptoSystem;
+
+    // generate a test agent id (HcS)
+    fn gen_id(crypto: &Box<dyn CryptoSystem>) -> String {
+        let enc = hcid::HcidEncoding::with_kind("hcs0").unwrap();
+        let mut key = crypto.buf_new_insecure(32);
+        crypto.randombytes_buf(&mut key).unwrap();
+        enc.encode(&key).unwrap()
+    }
+
+    // generate a test data address (HcA)
+    fn gen_data_addr(crypto: &Box<dyn CryptoSystem>) -> String {
+        let mut data = crypto.buf_new_insecure(32);
+        crypto.randombytes_buf(&mut data).unwrap();
+        let mut addr = crypto.buf_new_insecure(32);
+        crypto.hash_sha256(&mut addr, &data).unwrap();
+        let enc = hcid::HcidEncoding::with_kind("hca0").unwrap();
+        enc.encode(&addr).unwrap()
+    }
+
+    #[test]
+    fn it_should_safely_distribute_data() {
+        let crypto: Box<dyn CryptoSystem> =
+            Box::new(SodiumCryptoSystem::new().set_pwhash_interactive());
+
+        let mut nodes = Vec::new();
+
+        let mut min = 50;
+        let mut max = 50;
+        let mut count = 0;
+        let mut mean = 0.0;
+
+        // simulate a 10,000 node network, growing 10 nodes at a time
+        for _ in 0..1000 {
+            for _ in 0..10 {
+                nodes.push(anything_to_location(&crypto, &gen_id(&crypto)));
+            }
+
+            // simulate storing 100 bits of data in this network
+            for _ in 0..100 {
+                let data = anything_to_location(&crypto, &gen_data_addr(&crypto));
+
+                let mut store_count = 0;
+
+                // go through all the nodes
+                for agent_loc in nodes.iter() {
+                    if naive_sharding_should_store(*agent_loc, data, nodes.len() as u64) {
+                        store_count += 1;
+                    }
+                }
+
+                if nodes.len() < 50 {
+                    // if we have less than 50 nodes
+                    // make sure all nodes store all data
+                    assert_eq!(store_count, nodes.len());
+                } else {
+                    // if we have > 50 nodes,
+                    // assert that a reasonable number of nodes store the data
+                    assert!(store_count > 15);
```

Comment:
Bonus marks, what is the probability of a failure here? :P

Reply:
((n/50)-1)^n

Reply:
(1-(50/n))^n ?